File: modules/ISA-Matrix.py
"""
ISA-Matrix Module
Instruction Set Architecture (ISA) Matrix binding for Edge-AI-APP.
Provides ISA classification and mapping for CPU/GPU/NPU/FPGA execution targets.
"""

import sys
import json

# ISA classification matrix
ISA_MATRIX = {
    "x86_64": {
        "target": "CPU",
        "simd": ["SSE4.2", "AVX2", "AVX-512"],
        "precision": ["FP32", "FP64", "INT8"],
        "edge_support": False,
    },
    "ARM_Cortex-A": {
        "target": "CPU",
        "simd": ["NEON", "SVE"],
        "precision": ["FP32", "FP16", "INT8"],
        "edge_support": True,
    },
    "ARM_Cortex-M": {
        "target": "MCU",
        "simd": ["DSP", "MVE"],
        "precision": ["FP32", "INT8", "INT4"],
        "edge_support": True,
    },
    "RISC-V": {
        "target": "CPU/FPGA",
        "simd": ["RVV"],
        "precision": ["FP32", "FP16", "INT8"],
        "edge_support": True,
    },
    "CUDA_SM89": {
        "target": "GPU",
        "simd": ["Tensor Cores", "WMMA"],
        "precision": ["FP32", "FP16", "BF16", "INT8", "FP8"],
        "edge_support": False,
    },
    "NPU_Generic": {
        "target": "NPU",
        "simd": ["MAC Array"],
        "precision": ["INT8", "INT4", "FP16"],
        "edge_support": True,
    },
    "FPGA_LUT": {
        "target": "FPGA",
        "simd": ["DSP48", "Custom"],
        "precision": ["FP32", "FP16", "INT8", "Custom"],
        "edge_support": True,
    },
}


def build_isa_matrix():
    """Build and return the ISA compatibility matrix."""
    matrix = []
    for isa_name, props in ISA_MATRIX.items():
        row = {
            "ISA": isa_name,
            "Target": props["target"],
            "SIMD_Extensions": ", ".join(props["simd"]),
            "Precision_Formats": ", ".join(props["precision"]),
            "Edge_Capable": props["edge_support"],
        }
        matrix.append(row)
    return matrix


def print_matrix(matrix):
    """Print the ISA matrix in a readable table format."""
    header = f"{'ISA':<20} {'Target':<12} {'Edge':<6} {'Precision'}"
    print("=" * 80)
    print("ISA-Matrix: Instruction Set Architecture Binding Table")
    print("=" * 80)
    print(header)
    print("-" * 80)
    for row in matrix:
        print(
            f"{row['ISA']:<20} {row['Target']:<12} {str(row['Edge_Capable']):<6} {row['Precision_Formats']}"
        )
    print("=" * 80)


def select_isa_for_workload(workload_type: str) -> list:
    """Select suitable ISAs for a given workload type."""
    recommendations = []
    workload_map = {
        "inference": ["ARM_Cortex-A", "NPU_Generic", "FPGA_LUT", "CUDA_SM89"],
        "training": ["CUDA_SM89", "x86_64"],
        "edge": ["ARM_Cortex-M", "ARM_Cortex-A", "RISC-V", "NPU_Generic", "FPGA_LUT"],
        "fpga": ["FPGA_LUT", "RISC-V"],
    }
    targets = workload_map.get(workload_type.lower(), list(ISA_MATRIX.keys()))
    for isa in targets:
        if isa in ISA_MATRIX:
            recommendations.append({"ISA": isa, **ISA_MATRIX[isa]})
    return recommendations


def main():
    print("Initializing ISA-Matrix module...")
    matrix = build_isa_matrix()
    print_matrix(matrix)

    print("\nRecommended ISAs for edge inference:")
    edge_isas = select_isa_for_workload("edge")
    for item in edge_isas:
        print(f"  - {item['ISA']} ({item['target']})")

    # Export matrix as JSON for downstream pipeline consumption
    output = {"module": "ISA-Matrix", "version": "1.0.0", "matrix": matrix}
    print("\nISA-Matrix JSON output (summary):")
    print(json.dumps(output, indent=2)[:500] + "...")

    print("\nISA-Matrix module completed successfully.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/ISA-Matrix.c
/*
 * ISA-Matrix.c
 * Instruction Set Architecture Matrix binding for Edge-AI-APP.
 * Compiled to output/ISA-Matrix.bin by the CI/CD workflow.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NUM_ISA 7

typedef struct {
    const char *name;
    const char *target;
    const char *precision;
    int edge_capable;
} ISAEntry;

static const ISAEntry ISA_TABLE[NUM_ISA] = {
    {"x86_64",       "CPU",      "FP32/FP64/INT8",           0},
    {"ARM_Cortex-A", "CPU",      "FP32/FP16/INT8",           1},
    {"ARM_Cortex-M", "MCU",      "FP32/INT8/INT4",           1},
    {"RISC-V",       "CPU/FPGA", "FP32/FP16/INT8",           1},
    {"CUDA_SM89",    "GPU",      "FP32/FP16/BF16/INT8/FP8",  0},
    {"NPU_Generic",  "NPU",      "INT8/INT4/FP16",            1},
    {"FPGA_LUT",     "FPGA",     "FP32/FP16/INT8/Custom",    1},
};

void print_isa_table(void) {
    printf("%-20s %-12s %-6s %s\n", "ISA", "Target", "Edge", "Precision");
    printf("%-20s %-12s %-6s %s\n",
           "--------------------", "------------", "------",
           "----------------------------");
    for (int i = 0; i < NUM_ISA; i++) {
        printf("%-20s %-12s %-6s %s\n",
               ISA_TABLE[i].name,
               ISA_TABLE[i].target,
               ISA_TABLE[i].edge_capable ? "Yes" : "No",
               ISA_TABLE[i].precision);
    }
}

int main(int argc, char *argv[]) {
    printf("ISA-Matrix Binary v1.0.0\n");
    printf("========================\n\n");
    print_isa_table();
    printf("\nISA-Matrix binary executed successfully.\n");
    return 0;
}
